### PR TITLE
add irace

### PR DESCRIPTION
**Pull request by RaphaelS1**
* Adds TunerIrace
* Adds irace dependency
* Adds `irace_helpers.R` for helper functions to make `TunerIrace` a bit neater
* Adds tests for `TunerIrace`

Also changed the template very slightly to include all three constructor methods. If you dislike this let me know and I'll revert back.

### CODE REVIEW

File: R/irace_helpers.R
Comment (mb706) [inline R/irace_helpers.R]: I would always use `{` `}` in multi-line if statements.

File: R/irace_helpers.R
Comment (mb706) [inline R/irace_helpers.R]: Now my other comments were nitpicking, but I believe this is a serious error. The `instances` of irace are "challenges" to a given parameter configuration that this configurations is evaluated on. The results of these challenges are compared between different parameter configurations (in a pairwise test afair).

`instances` are not what we call a `TuningInstance`. Instead, they should be different resampling instances (or other things that are drawn at random from a population of which we are trying to get a configuration with best average performance: could be different tasks). I guess our `TuningInstance` the way we have it implemented currently doesn't work like that because it does not accept resampling instances as parameters. `TuningInstance` will probably have to be adjusted.

File: R/irace_helpers.R
Comment (mb706) [inline R/irace_helpers.R]: don't use `ifelse` unless you want to have vectorization; use `if () x else y`

Comment (mb706) [reply R/irace_helpers.R]: also you are checking whether the terminator is based on number of evals, but then relying on the number given in the paramset. shouldn't you use the number given in the terminator?

Comment (mb706) [reply R/irace_helpers.R]: But that is a moot point because using the terminator in such a way is broken. We expect the user to create new `Terminator` classes, e.g. to combine other terminators; the terminators should be handled as black boxes imho. (Unfortunately I guess it's the best we can do for now so)

File: R/TunerIrace.R
Comment (mb706) [inline R/TunerIrace.R]: you should probably overload the `assign_result` function and ask the irace result on what configuration it believes to be best. See my mattermost comments on that.

File: R/irace_helpers.R
Comment (mb706) [inline R/irace_helpers.R]: answer: irace does not do transformation, but `eval_batch()` does this for us so don't worry.

File: R/irace_helpers.R
Comment (mb706) [inline R/irace_helpers.R]: data.table probably has a more elegant way to do this

File: R/irace_helpers.R
Comment (mb706) [inline R/irace_helpers.R]: `as.data.table()` should be enough

Comment (RaphaelS1) [reply R/irace_helpers.R]: What do you mean by enough?

File: R/irace_helpers.R
Comment (RaphaelS1) [reply R/irace_helpers.R]: do you know the best vectorized alternative to `subset`?

File: R/irace_helpers.R
Comment (RaphaelS1) [reply R/irace_helpers.R]: ah yes I actually meant to remove this, but thanks for confirming!

File: R/TunerIrace.R
Comment (RaphaelS1) [reply R/TunerIrace.R]: Why get rid of "default" though? Just because this goes against the paradox convention of only labelling default if that's in the function? Or do you think user's should be forced to decide?

File: R/irace_helpers.R
Comment (RaphaelS1) [reply R/irace_helpers.R]: It has been tested. irace takes in everything in an expression (or deparsed?) form, so all numerics/non-numerics look identical but it then internally coerces back to required form

Comment (RaphaelS1) [reply R/irace_helpers.R]: > `||` instead of `|` would be nicer here

Where do you want `||` ? The use of `|` in the paste in line 32 is required by `irace`. It isn't stating "or" but that the condition is optional.

File: R/irace_helpers.R
Comment (RaphaelS1) [reply R/irace_helpers.R]: So this is exactly the part that I needed checking. The thing that threw me is actually the [mlr2 implementation]. You might be able to interpret this better than me but:

1. Within the individual experiment we have the [first instance always being selected](https://github.com/mlr-org/mlr/blob/1cb54c7e3750f2a20afc0740f368e64a3a5df76e/R/tuneIrace.R#L14-L15).
2. Then [new resample instances being created](https://github.com/mlr-org/mlr/blob/1cb54c7e3750f2a20afc0740f368e64a3a5df76e/R/tuneIrace.R#L29) up to the number specified in `n.instances`

I'm not familiar enough with the mlr2 or mlr3 implementations to know how creating multiple resamples from an instance in mlr2 differs from what is done in mlr3tuning. What we could do is replicate the `TuningInstance` in by `n.instances` if this creates different resampling schemes, or challenges as you put it.

File: R/irace_helpers.R
Comment (RaphaelS1) [reply R/irace_helpers.R]: > also you are checking whether the terminator is based on number of evals, but then relying on the number given in the paramset. shouldn't you use the number given in the terminator?

Sorry confused here, I am using the number given in the terminator... isn't that `instance$terminator$param_set$values$n_evals` ?

> But that is a moot point because using the terminator in such a way is broken. We expect the user to create new Terminator classes, e.g. to combine other terminators; the terminators should be handled as black boxes imho. (Unfortunately I guess it's the best we can do for now so)

Yes this is frustrating as this doesn't combine well with the package. The other problem is using different and confusing language. Terminator: "n_evals". irace: "n.instances" and "maxExperiments".

File: R/irace_helpers.R
Comment (RaphaelS1) [reply R/irace_helpers.R]: or I guess `join` instead

Comment (RaphaelS1) [reply R/irace_helpers.R]: I've used merged, not sure it is more elegant but I'll let you decide

File: R/irace_helpers.R
Comment (RaphaelS1) [reply R/irace_helpers.R]: Any need for `deparse` here though? This functions as expected

**Comment by berndbischl** [comment]
some small comments for now:
- i dislike that you are tagging every irace-control param with "irace". just to be able to add the meta-param "show.irace.output". can we somehow get around this?
- your coding style is in many place not mlr3-complaint (arrow, spaces, camel-case, etc). pls fix now
- you need to set 
        properties = "dependencies"
to tell the tuner that it can handle dependent params

**Comment by berndbischl** [comment]
- please unit-test paradox_to_irace separately
- mlr2 had a lot of unit tests for irace. copy most of them over if possible please

File: tests/testthat/test_TunerIrace.R
Comment (RaphaelS1) [inline tests/testthat/test_TunerIrace.R]: unsure if these tests are working as expected/testing the correct thing

File: tests/testthat/test_TunerIrace.R
Comment (RaphaelS1) [inline tests/testthat/test_TunerIrace.R]: unsure if these tests are working as expected/testing the correct thing

**Comment by RaphaelS1** [comment]
> i dislike that you are tagging every irace-control param with "irace". just to be able to add the meta-param "show.irace.output". can we somehow get around this?
your coding style is in many place not mlr3-complaint (arrow, spaces, camel-case, etc). pls fix now
you need to set
properties = "dependencies"
to tell the tuner that it can handle dependent params

Done, done, done.

> please unit-test paradox_to_irace separately
mlr2 had a lot of unit tests for irace. copy most of them over if possible please

Done - Need to check how dependencies are working though. I'm unsure if some tests working as expected. Also added a few dependencies to suggests that could be removed if needed by changing tested learners

File: R/irace_helpers.R
Comment (RaphaelS1) [reply R/irace_helpers.R]: > also has this been tested? it looks like `$rhs` should be quoted if it is non-numeric.

You were right, this is now fixed

File: R/TunerIrace.R
Comment (mb706) [reply R/TunerIrace.R]: Currently it is "just because" of the convention, but paradox may need to be modified in a way that assumes that users adhere to this convention, e.g. https://github.com/mlr-org/paradox/issues/265

File: R/irace_helpers.R
Comment (mb706) [reply R/irace_helpers.R]: you can probably do something with
```r
as.data.table(ps)[, ifelse(is.na(lower),
  <whatever paste collapse 'levels'>,
  paste0("(", lower, ",", upper, ")"))]
```

File: R/irace_helpers.R
Comment (mb706) [reply R/irace_helpers.R]: ah ok didn't know about the `"|"`

File: R/irace_helpers.R
Comment (mb706) [reply R/irace_helpers.R]: it is a bit cumbersome to essentially do the deparsing manually when in theory you should have an R function that does it for you, but you can go both ways.

File: R/irace_helpers.R
Comment (mb706) [reply R/irace_helpers.R]: > Sorry confused here, I am using the number given in the terminator... isn't that `instance$terminator$param_set$values$n_evals` ?

you are right, I misread that.

File: R/irace_helpers.R
Comment (RaphaelS1) [reply R/irace_helpers.R]: Do you know the best way to collapse lists with some elements characters and some not? 

```R
as.data.table(ps)[,ifelse(is.na(lower),
                            paste0("(",paste0(levels, collapse = ","),")"),
                            paste0("(",lower,",",upper,")"))]
```

Doesn't work as it collapses all `levels` and ignores the `ifelse, e.g. I got:

```
"c(TRUE, FALSE),NULL,c(\"lvl1\", \"lvl2\"),NULL" "(1,9)" 
```

File: R/TunerIrace.R
Comment (RaphaelS1) [reply R/TunerIrace.R]: I'm trying to do this but I'm running into a problem where even though `irace` is correctly calling each configuration with new parameters and evaluating these on each instance, for some reason the archived instance parameters do not align with `irace`. Which would be fine except `irace` only returns the optimal parameters but not the final performance cost

**Comment by RaphaelS1** [comment]
Shall we try to finish, merge and close this PR? @mb706 @berndbischl . Open questions seem to be about correctly setting `instances` to match mlr3tuning to irace design

**Comment by jakob-r** [comment]
We should first merge #240 and then I will probably have to adapt this PR to the API changes.

**Comment by RaphaelS1** [comment]
> We should first merge #240 and then I will probably have to adapt this PR to the API changes.

okay great, I've converted to a draft for now

**Comment by RaphaelS1** [comment]
@jakob-r after the merge from #240 (and probably many others since then), what needs to change in this PR?

**Comment by jakob-r** [comment]
Have you merged the master into your branch, check that tests are complete and running, mark this PR as ready for review and we will merge it if everyhting is fine.

**Comment by RaphaelS1** [comment]
Okay so pinging everyone who replied above many months ago: @jakob-r @be-marc @mb706 

I've merged in Master and got it working, however there seems to be a problem upstream causing warnings ('partial match') you can see it on all the builds. For most of the tests I can handle this with `suppressWarnings` but I can't do this on the primary autotest. The hacky fix would be to catch irace in the autotest and then wrap the public `optimize` method in `suppressWarnings`but that's up to you. I'm 99% sure the problem isn't our side because we don't use `length` as an argument in a package (grep), but let me know if wrong.

About the tuner itself, some things I'd appreciate double checking before merging, sorry I don't know the internals of mlr3tuning or irace well so a lot of my implementation is unfortunately guesswork:

1. Is the use of instance correct and concordant with the use in the irace packages? See conversations about this
2. Currently it's only set-up to work with `TerminatorEvals` and `TerminatorRunTime` as the parameters of these are passed to the `maxExperiments` and `maxTime` arguments in the irace scenario (see `make_scenario` in `irace_helpers.R`), i) do you agree with this?; ii) what does this mean for other terminators? i.e. will they just not work here or do we have to add them in some other way?
3. Can someone confirm that they agree that minimizing time will work as expected? See lines 43-52 in `test_TunerIrace.R` for an example
4. Follow-up from the conversation I had with @mb706 above regarding `assign_result`, does this still need updating or will the result automatically returned be the 'correct one'. Sorry I'm not very clear here but I'm also not sure at all about the internal of mlr3tuning so it's hard for me to know how the 'best' configuration is returned at the end of tuning (same with irace not sure how this works internally).

Thanks for the help with this

**Comment by mllg** [comment]
> I've merged in Master and got it working, however there seems to be a problem upstream causing warnings ('partial match') you can see it on all the builds. For most of the tests I can handle this with `suppressWarnings` but I can't do this on the primary autotest. The hacky fix would be to catch irace in the autotest and then wrap the public `optimize` method in `suppressWarnings`but that's up to you. I'm 99% sure the problem isn't our side because we don't use `length` as an argument in a package (grep), but let me know if wrong.

These checks are enabled by our `setup.R`, i.e. https://github.com/mlr-org/mlr3tuning/blob/master/tests/testthat/setup.R#L4.

You can temporarily disable them if required.

File: R/irace_helpers.R
Comment (be-marc) [inline R/irace_helpers.R:77]: > The objective of irace is to minimize the cost value returned by the target algorithm. If you wish to maximize, you can multiply the cost by -1 before returning it to irace.

(irace documentation)

Do you minimize somewhere else? If not you can use `scenario$instances[[1]]$objective_function(config)` which always minimizes.

Comment (RaphaelS1) [reply R/irace_helpers.R]: changed to `objective_function`

**Comment by be-marc** [comment]
@RaphaelS1 Hope you do not mind that I changed your code. `irace` does not fit into our `bbotk` / `mlr3tuning` framework like the other tuners.

> 1. Is the use of instance correct and concordant with the use in the irace packages? See conversations about this

No, instances in `irace` are resampling instances. We pass instantiated resampling instances in `scenario$instances` now. The user sets the number of instances via the `n_instances` parameter and internally the resampling is instantiated `n_instances` times. The `TuningInstanceSingleCrit` is now passed via `scenario$targetRunnerData`. 

> 2. Currently it's only set-up to work with TerminatorEvals and TerminatorRunTime as the parameters of these are passed to the maxExperiments and maxTime arguments in the irace scenario (see make_scenario in irace_helpers.R), i) do you agree with this?; ii) what does this mean for other terminators? i.e. will they just not work here or do we have to add them in some other way?

If the best configuration in `irace` is not based on the best performance in one resampling (Question 4), we should not terminate `irace` with our terminators. We should  `irace` terminate on its own so it can return the best configuration. 

@berndbischl You already mentioned in #169 that we might have a design issue. `TunerIrace` needs to change the train / test splits of `ObjectiveTuning$resampling`.

I solved this by making `ObjectiveTuning$resampling` an active binding.

```r
tuning_instance$objective$resampling = experiment$instance
cost = tuning_instance$eval_batch(config)
```

So `TunerIrace` changes the train / test split before executing `$eval_batch()`. 

Another option would be that `$eval_batch(config, resampling)` can also pass a resampling instance. We have to change the mlr3tuning code at multiple locations for this and copy/ modify methods that live in bbotk.

File: R/irace_helpers.R
Comment (be-marc) [reply R/irace_helpers.R]: Sorry but we have to use `$eval_batch()` again. I allows us to pass extra information like `id.instance` to the archive.

Comment (RaphaelS1) [reply R/irace_helpers.R]: Okay, change whatever needs changing!

**Comment by RaphaelS1** [comment]
> @RaphaelS1 Hope you do not mind that I changed your code. irace does not fit into our bbotk / mlr3tuning framework like the other tuners.

Please change anything you want, I really don't know the internals of mlr3tuning too well and I know nothing about bbotk. This initial PR was more of a starting point that I hope provides some help!

> No, instances in irace are resampling instances. We pass instantiated resampling instances in scenario$instances now.

Yes I think this changed considerably after I opened the initial PR

**Comment by be-marc** [comment]
> This initial PR was more of a starting point that I hope provides some help!

Yes your work helps a lot!

> Follow-up from the conversation I had with @mb706 above regarding assign_result, does this still need updating or will the result automatically returned be the 'correct one'. Sorry I'm not very clear here but I'm also not sure at all about the internal of mlr3tuning so it's hard for me to know how the 'best' configuration is returned at the end of tuning (same with irace not sure how this works internally).

`irace::irace` returns the elites of the final race as the result. We select the best performing one in `.assign_result`. The reported performance is the average performance of all resampling iterations. The log of `irace::irace` is returning the same number.

**Comment by be-marc** [comment]
We have to remove `show.irace.output`. It also filters our logging messages. We do not have a solution for this. `regr.km` is also spaming the console.

**Comment by be-marc** [comment]
> Can someone confirm that they agree that minimizing time will work as expected? See lines 43-52 in test_TunerIrace.R for an example

The unit tests that uses `TerminatorRunTime` fail sometimes i.e. `irace::irace` does return a best configuration. 

```r
ps = ParamSet$new(params = list(
  ParamDbl$new("cp", lower = 0.001, upper = 0.1),
  ParamDbl$new("minsplit", lower = 1, upper = 10)
))
ps$trafo = function(x, param_set) {
  x$minsplit = as.integer(round(x$minsplit))
  return(x)
}
inst = TuningInstanceSingleCrit$new(
  tsk("iris"), lrn("classif.rpart"), rsmp("holdout"),
  msr("classif.ce"), trm("run_time", secs = 24), ps)
tt = tnr("irace")
tt$optimize(inst)
expect_double(inst$archive$best()$cp)
```

```r
ps = ParamSet$new(params = list(
  ParamDbl$new("cp", lower = 0.001, upper = 0.1)
))
inst = TuningInstanceSingleCrit$new(
  tsk("iris"), lrn("classif.rpart"), rsmp("holdout"),
  msr("classif.ce"), trm("run_time", secs = 20), ps)
tt = tnr("irace", capping = 1, boundMax = 1, cappingType = "best", boundType = "instance")
tt$optimize(inst)
expect_double(inst$archive$best()$cp)
```

**Comment by be-marc** [comment]
I added a new unit test with `TerminatorRunTime` that seems to run stable.  The PR is ready for a review.

File: R/ObjectiveTuning.R
Comment (jakob-r) [inline R/ObjectiveTuning.R:96]: Is this change necessary for irace or just a general "improvement"?

File: R/ObjectiveTuning.R
Comment (jakob-r) [inline R/ObjectiveTuning.R:47]: The assertion would be acchived by the active binding, right?

File: R/ObjectiveTuning.R
Comment (jakob-r) [reply R/ObjectiveTuning.R:96]: Is it documented that the resampling is instantiated once the objective is created? This is kind of a havy change as now the function is not as "random" anymore.

File: R/TunerIrace.R
Comment (jakob-r) [inline R/TunerIrace.R:108]: As defined in the objective the resampling is already instantiated. Why doing it here again? Do we want the same resampling splits in all scenarios for paired testing or do we specifically want the opposite? Anyhow, those things should be commented!

File: R/TunerIrace.R
Comment (jakob-r) [inline R/TunerIrace.R:118]: `inherits()`?

File: R/TunerIrace.R
Comment (jakob-r) [inline R/TunerIrace.R:102]: We should also stop if the terminator is not supported.

File: R/TunerIrace.R
Comment (jakob-r) [inline R/TunerIrace.R:137]: `get("id_configuration")` so you don't have to do this NULL thing

File: R/TunerIrace.R
Comment (jakob-r) [reply R/TunerIrace.R:145]: Is this resolved? Looks like it. Please mark as solved then.

File: R/irace_helpers.R
Comment (jakob-r) [inline R/irace_helpers.R:60]: `target_runner`?

File: R/irace_helpers.R
Comment (jakob-r) [inline R/irace_helpers.R:84]: If I remember correctly `as.numeric(t1 - t0)` is dangerous because it can be either 1 for seconds or 1 for minutes. There is a away to define the unit.

**Review by jakob-r — CHANGES_REQUESTED** [review]
Thanks for the effort. Looks good for the most part. Have we discussed it this would be worth an extra package?

**Comment by jakob-r** [comment]
> ~I added a new unit test with `TerminatorRunTime` that seems to run stable.~ The PR is ready for a review.
> The unit test with `TerminatorRunTime` is still not stable. It rarely fails and is difficult to reproduce. If `irace::irace` does not return a result, we throw the error message `irace::irace did not return a result. The evaluated configurations are still accessible through the archive.` The unit test wraps `tuner$optimize(instance)` in a try statement to avoid failing tests. We just check if the archive has evaluated configurations. Does anyone have a better idea?

Does it fail bc. irace is taking a tiny bit longer than the terminator is allowing?

File: R/ObjectiveTuning.R
Comment (be-marc) [reply R/ObjectiveTuning.R:96]: If we leave `resample` a field, a tuner could change it to an uninstantiated resampling and each objective function call would be evaluated on different train/ test splits. Currently, we instantiate the resampling during the initialization. It is a general improvement since we have to pass already instantiated resamplings to irace (see comment bellow). We could discus a different way to do this in #169.

File: R/TunerIrace.R
Comment (be-marc) [reply R/TunerIrace.R:108]: We want to give `irace` n different resampling train/ test splits. `irace` then decides which configurations are evaluated on which resamplings. We could also pass `1:n` here and the number is the seed of the train/ test split. But I think this is more complicated.

Comment (be-marc) [reply R/TunerIrace.R:108]: See `target_runner()`. `irace` gives us the  configuration `experiment$configuration` and the resampling `experiment$instance` on which have to evaluate the configuration.

File: R/TunerIrace.R
Comment (be-marc) [reply R/TunerIrace.R:102]: Added a stop.

File: R/TunerIrace.R
Comment (be-marc) [reply R/TunerIrace.R:137]: I am always confused when `get()` works...

File: R/irace_helpers.R
Comment (be-marc) [reply R/irace_helpers.R:84]: `as.numeric(difftime(t1, t0, units = "secs"))`

**Comment by be-marc** [comment]
Our documentation stated 

> Uninstantiated resamplings are instantiated during construction so that all configurations are evaluated on the same data splits.

I added

> If a new resampling is passed, it is instantiated with new data splits. Already instantiated resamplings are kept unchanged.

**Comment by be-marc** [comment]
> Does it fail bc. irace is taking a tiny bit longer than the terminator is allowing?

It does not fail often. We substract 5 seconds now. Maybe it helps.

File: R/TunerIrace.R
Comment (jakob-r) [inline R/TunerIrace.R:113]:
```suggestion
      ri = replicate(pv$n_instances, {
```
did the other even work?

**Review by jakob-r — CHANGES_REQUESTED** [review]

File: R/TunerIrace.R
Comment (be-marc) [reply R/TunerIrace.R:113]: Thanks. No I pushed that by mistake...

**Comment by be-marc** [comment]
Closes #63

File: R/irace_helpers.R
Comment (mb706) [reply R/irace_helpers.R:27]: add `by = id` and use `levels[[1]]`:
```r
as.data.table(ps(x = p_dbl(1, 4), y = p_lgl()))[,
  ifelse(is.na(lower),
    paste0("(",paste0(levels[[1]], collapse = ","),")"),
    paste0("(",lower,",",upper,")")),
  by = "id"]$V1
#> [1] "(1,4)"        "(TRUE,FALSE)"
```

File: tests/testthat/setup.R
Comment (jakob-r) [inline tests/testthat/setup.R:3]: Which one?

**Review by jakob-r — APPROVED** [review]

**Comment by jakob-r** [comment]
I think this can be merged. Any objections?

File: tests/testthat/setup.R
Comment (be-marc) [reply tests/testthat/setup.R:3]: `irace` uses `rep` with `length` instead of `length.out`

Comment (be-marc) [reply tests/testthat/setup.R:3]: Reported upstream https://github.com/MLopez-Ibanez/irace/issues/7.

Comment (MLopez-Ibanez) [reply tests/testthat/setup.R:3]: For what is worth, this issue is now fixed upstream so you should be able to re-enable the warnings. We, the irace team (which is basically 2 people doing both teaching and research), are happy to receive issues and pull requests for fixes like these but also for changes that simplify integration with mlr (or other packages).

**Pull request merged**
